base/shell.jl

```diff
@@ -255,60 +255,91 @@ shell_escape_posixly(args::AbstractString...) =
     sprint(print_shell_escaped_posixly, args...)
 
 
-function print_shell_escaped_winsomely(io::IO, args::AbstractString...)
-    first = true
-    for arg in args
-        first || write(io, ' ')
-        first = false
-        # Quote any arg that contains a whitespace (' ' or '\t') or a double quote mark '"'.
-        # It's also valid to quote an arg with just a whitespace,
-        # but the following may be 'safer', and both implementations are valid anyways.
-        quotes = any(c -> c in (' ', '\t', '"'), arg) || isempty(arg)
-        quotes && write(io, '"')
-        backslashes = 0
-        for c in arg
-            if c == '\\'
-                backslashes += 1
-            else
-                # escape all backslashes and the following double quote
-                c == '"' && (backslashes = backslashes * 2 + 1)
-                for j = 1:backslashes
-                    # backslashes aren't special here
-                    write(io, '\\')
-                end
-                backslashes = 0
-                write(io, c)
+"""
+    shell_escape_wincmd(s::AbstractString)
+    shell_escape_wincmd(io, s::AbstractString)
+
+The unexported `shell_escape_wincmd` function escapes the meta
+characters `()!^<>&|` processed by the Windows `cmd.exe` shell: it
+places a `^` in front of any metacharacter that follows an even number
+of quotation marks on the command line.
+
+The percent sign (`%`) is not escaped, therefore shell variable
+references (like `%USER%`) will still be substituted by `cmd.exe`.
```

Comment:
This is not right: this is what `shell_escape_wincmd` exists to prevent.
```suggestion
```

Reply:
It's not so easy! The problem is that CMD.EXE performs variable substitution at the very beginning, _before_ processing the ^ and " escaping characters. See the following experiment to demonstrate this:
```
C:\Users\mgk25>echo %USER%            
dc\mgk25                     
                             
C:\Users\mgk25>echo ^%USER% 
dc\mgk25                     
                             
C:\Users\mgk25>echo ^%USER^% 
%USER%                       
                             
C:\Users\mgk25>set USER^^=foo

C:\Users\mgk25>echo ^%USER^%  
foo                          

```
So if you try to escape % with ^%, you may in fact just be adding a ^ to the variable name that will be substituted. Yes, I know its crazy, but Windows always was an acquired taste ...

I believe it is impossible for `shell_escape_wincmd` to escape % effectively. I can reformulate the documentation to clarify that passing through % was not our choice, but was done to clarify that escaping it would not have been effective anyway.

Reply:
Oh, wow. Yeah, I completely missed that. (I doubt I'm alone too, since it means 8.10 in http://daviddeley.com/autohotkey/parameters/parameters.htm#WINCRULES is similarly wrong). It's very unexpected to realize that `%` is _not_ a cmd shell metacharacter, so `^%` is no different from say `^A`. It's instead a half separate pass of the parser.

Following the someone more detailed guide at https://stackoverflow.com/questions/4094699/how-does-the-windows-command-interpreter-cmd-exe-parse-scripts/7970912#7970912, I realized there's a workaround for rules that is valid for your experiment above. That is, an almost-working escape sequence for `%` is `%^\n`, as can be seen in the below example:

```
julia> run(pipeline(`cmd /Q /K set PROMPT=\?`, stdin=IOBuffer("echo %^\nOS%^\n\n\n")))

?More? More? More? %OS%


?Process(`cmd /Q /K set 'PROMPT=?'`, ProcessExited(0))
```

Note that passing this to `/C` fails though because of rule 1.05: "the remainder of the line from the <LF> onward" is dropped (ignored). Note that we are gaining an extra newline on the output too, since we're running afoul of the quoting rules for `\n` (it's `^\n\n`) to handle the trailing `%`—there would need to be an exception to the rule for when `%` is at the end of the line already.

Also, perhaps we should mention that the user is advised to normally pass the S flag as such: `/S /C " $the $command $line "` to avoid losing their final `"` mark in `$line` contents under very rare circumstances.

Reply:
This is starting to get rather esoteric. I actually want to pass through % unescaped, because my use case is building a command line to be sent with ssh to _another_ Windows server, where it can be perfectly useful to have variables such as %APPDATA% substituted remotely (as I don't have local access to them). Also, in the ssh use case, I am not calling CMD.EXE, so I can't pass any flags to it, the remote sshd is doing that.

If someone really needs a variant of this function that also goes to such great lengths to escape `%`, then I would be happy to draft a separate function `shell_escape_wincmd2` that has that property. However, the algorithm used will be completely different, because in order to escape every `%` with your hack, I can no longer pass through pairs of `"` unescaped, as they too might contain a `%`. Escaping every `%` means also escaping every `"`, and I'd rather not do the latter as it makes all the most common cases look very odd.

Reply:
Yeah, that's really abysmal behavior/design by CMD since it means you can't easily have one script call another (say, on another machine). I guess we could also attempt to hack it with `ENV["PCTESC"]="%"; replace(cmd, "%" => "%PCTESC%")`, but that seems pretty brittle too.